### PR TITLE
Fix bugs (ping/encoder/dispose_method/position) in webp coder

### PR DESCRIPTION
**Pull request by joonsung90**
Hi.
I've found some bugs in the webp coder (both decoder & encoder).
Please refer to the below for more details.

1. "Image ping" cannot extract the number of scenes correctly. When I aim to get properties of an animated webp image via “MagickPingImageBlob” function, a wand object returned from the function doesn’t contain the exact number of scenes. (MagickGetNumberImages always returns 1). So, I checked the webp decoder part, and I found the decoder only loads entire image frames when the ping flag doesn’t set. Therefore, I suggest the decoder load full frames whether a given request is ping type or not.

2. Here, the current decoder assumes the first page doesn’t have positional information. The webp library correctly provides the size of virtual its virtual canvas by parsing headers; however, the current encoder doesn’t use it. Please check the attached webp file. if one tries to convert that webp file into gif file (“convert test.webp test.gif”) Each position of a frame in images is totally jumbled.

3. Also, the current decoder doesn’t reflect the dispose method of the original webp image correctly. So I suggest setting image->dispose by checking iter.dispose_method value.

4. There’s a bug in the encoding function (i.e., WriteWEBPImage). The function requires three arguments (image_info, image, exception); however, this function is called with two arguments (except exception).

Again, Thanks for your efforts to maintain ImageMagick library 👍

### CODE REVIEW

**Comment by dlemstra** [comment]
Thanks for your contribution and I will take a look at it this weekend. I noticed that you used another commit email than the one you use in your profile. Feel free to ammend your commit to use the same email.

**Comment by joonsung90** [comment]
Thanks for letting me know the discrepancy in user name & email. :)

p.s. Would you be willing to help us with IM7 once this is merged?
=> Yes, I am willing to :)

**Comment by dlemstra** [comment]
It looks like you got a merge commit instead of a rebase. Can you please only include commit "c59f1ccc55a48e58cd2e29da7439c5c1262d4ef9" in this PR?

File: coders/webp.c [hunk 1]
```diff
@@ -833,7 +878,7 @@ static MagickBooleanType WriteAnimatedWEBPImage(const ImageInfo *image_info,
 #endif
 
 static MagickBooleanType WriteWEBPImage(const ImageInfo *image_info,
-  Image *image,ExceptionInfo * exception)
+  Image *image)
```
Comment (dlemstra) [inline coders/webp.c]: That is a really good catch.

File: coders/webp.c [hunk 2]
```diff
@@ -249,14 +259,32 @@ static int ReadSingleWEBPImage(Image *image,const uint8_t *stream,
...
+    image->page.x = 0;
+    image->page.y = 0;
+  } else {
+    x_offset = 0;
+    y_offset = 0;
+  }
   webp_status = FillBasicWEBPInfo(image,stream,length,configure);
+  image_width = image->columns;
+  image_height = image->rows;
+  if (is_first) {
```
Comment (dlemstra) [inline coders/webp.c]: We have a different coding convention for `if` statements. Could you change it to this:

```c
if (is_first)
  {
    image->columns = canvas_width;
    image->rows = canvas_height;
  }
```

There are more statements in this PR so please change all of them.

File: coders/webp.c [hunk 3]
```diff
@@ -232,11 +232,21 @@ static int FillBasicWEBPInfo(Image *image,const uint8_t *stream,size_t length,
 }
 
 static int ReadSingleWEBPImage(Image *image,const uint8_t *stream,
-  size_t length,WebPDecoderConfig *configure,ExceptionInfo *exception)
+  size_t length,WebPDecoderConfig *configure,ExceptionInfo *exception,MagickBooleanType is_first)
```
Comment (dlemstra) [inline coders/webp.c]: We use a line limit of 80 characters in our project. Could you move this to the next line?

File: coders/webp.c [hunk 4]
```diff
@@ -232,11 +232,21 @@ static int FillBasicWEBPInfo(Image *image,const uint8_t *stream,size_t length,
...
 {
   int
     webp_status;
 
+  size_t
```
Comment (dlemstra) [inline coders/webp.c]: Thanks for ordering the properties but we also order the type. Could you move this below `register unsigned char` and there is also already a `ssize_t` block.

File: coders/webp.c [hunk 5]
```diff
@@ -249,14 +259,32 @@ static int ReadSingleWEBPImage(Image *image,const uint8_t *stream,
   MagickBooleanType
     status;
 
+  if (is_first) {
+    canvas_width = image->columns;
```
Comment (dlemstra) [inline coders/webp.c]: Sorry for some more OCD code style comments but could you remove the spaces and change it to `canvas_width=image->columns`? This file already has a mixed convention but I would prefer if we don't introduce new ones. Feel free to reach out if you are in doubt about a convention.

File: coders/webp.c [hunk 6]
```diff
@@ -249,14 +259,32 @@ static int ReadSingleWEBPImage(Image *image,const uint8_t *stream,
...
+  }
+
   if(webp_status != VP8_STATUS_OK)
     return(webp_status);
 
   if (IsWEBPImageLossless(stream,length) != MagickFalse)
     image->quality=100;
 
-  webp_status=WebPDecode(stream,length, configure);
+  webp_status=WebPDecode(stream,length,configure);
```
Comment (dlemstra) [inline coders/webp.c]: Thanks 👍

File: coders/webp.c [hunk 7]
```diff
@@ -371,28 +411,40 @@ static int ReadAnimatedWEBPImage(const ImageInfo *image_info,Image *image,
...
+      }
       if(webp_status != VP8_STATUS_OK)
         break;
 
+      image->page.width = canvas_width;
+      image->page.height = canvas_height;
       image->ticks_per_second=100;
       image->delay=iter.duration/10;
+      if (iter.dispose_method == WEBP_MUX_DISPOSE_BACKGROUND) {
+        image->dispose = BackgroundDispose;
```
Comment (dlemstra) [inline coders/webp.c]: Good addition.

**Comment by dlemstra** [comment]
Could you include the `test.webp` in your initial post? You might need to zip it if you cannot drag the file into your message.

**Comment by joonsung90** [comment]
Hi.
I fixed the codes to follow the programming convention. Also, I applied force-push to remove some dirty merge messages.

[test.webp.zip](https://github.com/ImageMagick/ImageMagick6/files/3937755/test.webp.zip)

In addition, I attached the 'test.webp' image. (I missed to upload it. Sorry)

Thanks.

**Comment by joonsung90** [comment]
Sorry, I missed the indent in if/else code block. I've fixed it.

**Comment by dlemstra** [comment]
I am not sure which block you are talking about but this looks good to me. Thanks for the extra effort

**Comment by dlemstra** [comment]
Do you need some help to get this into the IM7 code @joonsung90? Or did you just not have time yet to do this?

**Comment by joonsung90** [comment]
Hi.

Sorry for the delay. I recently have not had much time due to my company
affairs.
I will begin this work as soon as possible.

Thanks.
Joonsung Kim.

2019년 12월 17일 (화) 오전 4:30, Dirk Lemstra <notifications@github.com>님이 작성:

**Pull request merged**